File: Blind75/3. Longest Substring Without Repeating Characters/3. Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    //time complexity O(n)
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        int left = 0;
        int right = 0;
        int res = 0;
        unordered_map<char, int> freq;
        while(right<n){
            freq[s[right]]++;
            while(freq[s[right]]>1){
                freq[s[left]]--;
                left++;
            }
            res = max(res, right-left+1);
            right++;
        }
        return res;
    }
};

File: Blind75/3. Longest Substring Without Repeating Characters/3. Longest Substring Without Repeating Characters.cs
public class Solution {
    public int LengthOfLongestSubstring(string s) {
        Dictionary<char, int> dect = new Dictionary<char,int>();
        int res = 0;
        int n = s.Length;
        int left = 0;
        int right = 0;

        while(right<n){
            if(dect.ContainsKey(s[right])){
                dect[s[right]]++;
            }else{
                dect.Add(s[right],1);
            }
            while(dect[s[right]]>1){
                dect[s[left]]--;
                left++;
            }
            res = Math.Max(res, right - left+1);
            right++;
        }
        return res;
    }
}
